### PR TITLE
005 expected kainos competencies per band

### PR DESCRIPTION
**Pull request by TomaszLaski**

### CODE REVIEW

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 1]
```diff
@@ -51,4 +56,17 @@ public Response getJobSpecification(@PathParam("id") int role_id)
...
         }
     }
+    @GET
+    @Path("/band-competency/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getBandCompetency(@PathParam("id") int role_id)
+    {
+        try {
+            return Response.status(HttpStatus.OK_200).entity(competencyService.getAllCompetencyPerBandLvl(role_id)).build();
+        } catch (CompetencyPerBandLvlNotExistException e){
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: you can re-use role does not exist here

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.kainos.ea.model.CompetencyPerBandLvl;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CompetencyPerBandLvlDao {
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: CompetencyDao is a better name

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 3]
```diff
@@ -51,4 +56,17 @@ public Response getJobSpecification(@PathParam("id") int role_id)
             return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
         }
     }
+    @GET
+    @Path("/band-competency/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getBandCompetency(@PathParam("id") int role_id)
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: you shouldn't take in role id when competencies are related to band, update to take in band id

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 4]
```diff
@@ -51,4 +56,17 @@ public Response getJobSpecification(@PathParam("id") int role_id)
             return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
         }
     }
+    @GET
+    @Path("/band-competency/{id}")
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: this should be /competencies/{id}

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public CompetencyPerBandLvlDao(){
+    }
+    public List<CompetencyPerBandLvl> getAllCompetencyPerBandLvl(Connection c, int role_id) throws SQLException, CompetencyPerBandLvlNotExistException {
+
+        try{
+            Statement st = c.createStatement();
+            ResultSet rs = st.executeQuery("SELECT DISTINCT temp.id, band_level, band_name, competency_name FROM (\n" +
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: by taking in the band id in the URL this query should be much simpler

File: src/main/java/com/kainos/ea/model/CompetencyPerBandLvl.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class CompetencyPerBandLvl {
+    private int bandLvl;
+    private String bandName;
+    private String competencyName;
+
+
+    @JsonCreator
+    public CompetencyPerBandLvl(
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/model/CompetencyPerBandLvl.java]: BandCompetencies I believe would be a better name

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                    "          JOIN competency_band ON band.id = competency_band.band_id) AS temp\n" +
+                    "      JOIN competency ON competency_id=competency.id\n" +
+                    "      WHERE temp.id = " + role_id );
+            List<CompetencyPerBandLvl> competencyPerBandLvl = new ArrayList<>();
+
+            while (rs.next()) {
+                CompetencyPerBandLvl competencyPerBandLvls = new CompetencyPerBandLvl(
+                        rs.getInt("band_level"),
+                        rs.getString("band_name"),
+                        rs.getString("competency_name"));
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: the band name and band level are always going to be the same here right? so you model should have a band level, band name and a list of competency name so the fields are not repeated

File: src/test/java/com/kainos/ea/integration/WebServiceIntegrationTest.java [hunk 8]
```diff
@@ -74,4 +75,19 @@ void getJobRoles_shouldReturnListOfJobRoles_withFieldsNotNull() {
         assertNotNull(JobRoleList.get(0).getCapability_name());
     }
 
+    @Test
+    void getCompetencyPerBand_shouldReturnListOfCompetencies_withFieldsNotNull() {
```
Comment (shaunganley) [inline src/test/java/com/kainos/ea/integration/WebServiceIntegrationTest.java]: need an integration test that returns a 404

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java (hunk 7 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: done

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java (hunk 5 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: done

File: src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java (hunk 2 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/dao/CompetencyPerBandLvlDao.java]: done

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 4 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/controller/WebService.java]: done

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 3 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/controller/WebService.java]: done

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 1 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/controller/WebService.java]: done

File: src/main/java/com/kainos/ea/model/CompetencyPerBandLvl.java (hunk 6 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/model/CompetencyPerBandLvl.java]: done

File: src/test/java/com/kainos/ea/integration/WebServiceIntegrationTest.java (hunk 8 above)
Comment (TomaszLaski) [reply src/test/java/com/kainos/ea/integration/WebServiceIntegrationTest.java]: done

**Review by cdrainkainos — APPROVED** [review]
looks good

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 1 above)
Comment (shaunganley) [reply src/main/java/com/kainos/ea/controller/WebService.java]: This should be BandNotExistException

File: src/main/java/com/kainos/ea/dao/CompetencyDao.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CompetencyDao {
+    public CompetencyDao(){
+    }
+    public List<BandCompetencies> getAllCompetencyPerBandLvl(Connection c, int role_id) throws SQLException, CompetencyPerBandLvlNotExistException {
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/dao/CompetencyDao.java]: parameter name wrong

File: src/main/java/com/kainos/ea/dao/CompetencyDao.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                bandName= rs.getString("band_name");
+                String competencyBandName= (rs.getString("competency_name"));
+                competencyBandNames.add(competencyBandName);
+            }
+
+                BandCompetencies bandLvls = new BandCompetencies(
+                        bandLevel,
+                        bandName,
+                        competencyBandNames);
+                bandCompetencies.add(bandLvls);
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/dao/CompetencyDao.java]: I don't think this method should be return a list. I think it should be one object that has the band level, band name and a list of competencies?

File: src/main/java/com/kainos/ea/service/CompetencyService.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+public class CompetencyService {
+    public CompetencyDao competencyDao;
+    public DatabaseConnection databaseConnector;
+
+    public CompetencyService(CompetencyDao competencyDao, DatabaseConnection databaseConnector) {
+        this.competencyDao = competencyDao;
+        this.databaseConnector = databaseConnector;
+    }
+
+    public List<BandCompetencies> getAllCompetencyPerBandLvl(int role_id) throws SQLException, DatabaseConnectionException, IOException, CompetencyPerBandLvlNotExistException {
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/service/CompetencyService.java]: param name wrong

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 1 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/controller/WebService.java]: done

File: src/main/java/com/kainos/ea/dao/CompetencyDao.java (hunk 9 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/dao/CompetencyDao.java]: done

File: src/main/java/com/kainos/ea/dao/CompetencyDao.java (hunk 10 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/dao/CompetencyDao.java]: done

File: src/main/java/com/kainos/ea/service/CompetencyService.java (hunk 11 above)
Comment (TomaszLaski) [reply src/main/java/com/kainos/ea/service/CompetencyService.java]: done

**Review by shaunganley — APPROVED** [review]

**Pull request merged**